File: base/my_av.py
#!/usr/bin/python3

import re
try:
    from difflib import SequenceMatcher
except Exception as e:
    pass

def parse_url(url):
    path, query, fragment = '', '', ''
    if "://" in url:
        protocol, url = url.split("://")

    # geth the host
    host = url.split("/")[0]

    # get the fragment if existent
    if '#' in url:
        fragment = url.split("#")[1]

    # get the query if existent
    if '?' in url:
        query = url.split("?")[1].replace(f'#{fragment}', '')

    path = url.replace(host, '').replace(f'?{query}', '')

    return host, path, query, fragment


def is_malicious(host, path):
    # this function checks if the url contains any signs of malware

    # check for common file extensions
    if '.' in path:
        extensions = ['exe', 'bin', 'sh', 'pl']
        extension = path.split('.')[-1]
        if extension in extensions:
            return 1

    main_domain = host.split('.')[-2]
    # if known goood hosts, not malicious
    if main_domain in whitelist:
        return 0

    # if hostname is too similar but not the same with a whitelisted domain,
    # probbly malicious
    for good_host in whitelist:
        ratio = 0
        try:
            ratio = SequenceMatcher(None, good_host, main_domain).ratio()
        except Exception:
            pass
        if ratio > 0.7:
            return 1
        # also if whitelisted domain is included in hostname
        if good_host in main_domain and main_domain != good_host:
            return 1

    # if hostname is too long, may be malicious
    if len(host) > 31:
        return 1

    # if too many numbers may be a malicious ip
    no_numbers = 0
    for char in charset:
        if char in host:
            no_numbers += 1
    if no_numbers >= 0.1 * len(host):
        return 1

    # if connecting to a specific port or credentials, maybe malicious
    if ':' in host or '@' in host:
        return 1

    # if double com extension may be a junk url ,so malicious
    host = host + '/'
    if len(re.findall(r"([^\w]+)com([^\w]+|/)", host)) > 1:
        return 1
    host = host[:-1]

    # check for bad chars
    if '~' in path and '.htm' not in path:
        return 1

    if 'secur' in path or 'paypal' in path or 'wp-admin' in path:
        return 1

    return 0


# some useful data
charset = "1234567890"
whitelist = ['google', 'facebook', 'googlegroups', 'paypal', 'twitter', 'bing',
             '123people', 'whatsapp', 'bdnews24']

# load the list of known malicious hosts and strip newlines
with open("../data/url_dataset/domains_database", 'r') as f:
    malicious_domains = f.readlines()
    malicious_domains = [i.rstrip() for i in malicious_domains]

# now read the list of urls to analyze
urls_file = open("../data/url_dataset/urls.in", 'r')
predictions_file = open("urls-predictions.out", 'w')

# for each url, analyze
url = urls_file.readline().rstrip()
while url != '' and url:
    malicious = 0
    host, path, query, fragment = parse_url(url)

    # if host is known to be malicious, flag it
    for mal_domain in malicious_domains:
        if mal_domain in host:
            malicious = 1

    if is_malicious(host, path):
        malicious = 1

    # now output into the file
    predictions_file.write(f"{malicious}\n")

    # read each line and parse it
    url = urls_file.readline().rstrip()

urls_file.close()
predictions_file.close()

# HERE STARTS TASK 2

try:
    import time
    from datetime import timedelta
except Exception as e:
    pass

# some macros for field numbers
flow_duration = 4
flow_payload_avg = 16
src_ip_field = 0
dst_ip_field = 2

# cryptominers seem to have this info
cryptominer_payload_avg = 40.0

# this is the list with evil ips
evil_ips = ['']


def parse_time(duration):
    days, duration = duration.split(' days ')
    days = int(days) * 86400
    try:
        t_struct = time.strptime(duration, '%H:%M:%S.%f')
        seconds = timedelta(hours=t_struct.tm_hour, minutes=t_struct.tm_min,
                            seconds=t_struct.tm_sec).total_seconds()
        miliseconds = 0
        if '.' in duration:
            miliseconds = float(f"0.{duration.split('.')[1]}")
    except Exception as e:
        return 0
    # return total number of seconds
    return float(days) + seconds + miliseconds


def parse_traffic(packet):
    # get the fields we need for analyzing
    duration = packet.split(',')[flow_duration]
    payload_avg = packet.split(',')[flow_payload_avg]
    src_ip = packet.split(',')[src_ip_field]
    dst_ip = packet.split(',')[dst_ip_field]

    return parse_time(duration), float(payload_avg), src_ip, dst_ip


def is_malicious_traffic(packet):
    duration, payload_avg, src_ip, dst_ip = parse_traffic(packet)

    # if payload is 0, not malicious
    if payload_avg == 0.0:
        return 0

    # if broadcast, not malicious
    if '255.255.255.255' in dst_ip:
        return 0

    # check if IP is known to be bad
    if src_ip in evil_ips:
        return 1

    # if duration is bigger than one second, probably malicious
    if duration > 1.0:
        evil_ips.append(src_ip)
        return 1

    if payload_avg == cryptominer_payload_avg:
        evil_ips.append(src_ip)
        return 1

    return 0


# now read the list of traffic packets to analyze
traffic_file = open("../data/network_dataset/traffic.in", 'r')
predictions_file = open("traffic-predictions.out", 'w')

# first line is junk
packet = traffic_file.readline()
packet = traffic_file.readline().rstrip()

while packet != '' and packet:
    malicious = 0

    if is_malicious_traffic(packet):
        malicious = 1

    # now output into the file
    predictions_file.write(f"{malicious}\n")

    # read each line and parse it
    packet = traffic_file.readline().rstrip()

traffic_file.close()
predictions_file.close()

File: base/my_av.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAXLEN 1000
#define SIZE 1000
#define IP_MAXLEN 16

// some macros for usual time numers
#define DAY_SECONDS 86400
#define HOUR_SECONDS 3600
#define MINUTE_SECONDS 60

// some macros for field numbers
#define FLOW_DURATION_DAYS 4
#define FLOW_DURATION_HOURS 6
#define FLOW_DURATION_MINS 7
#define FLOW_DURATION_SECS 8
#define FLOW_PAYLOAD_AVG 20
#define SRC_IP 0
#define DST_IP 2

void split(char string[], char sep, char result[])
{
	// this function takes the string before a separator if existent
	char sep_arr[] = { sep, '\0'}, tmp[MAXLEN];
	strcpy(tmp, string);
	if (strchr(tmp, sep))
		strcpy(result, strtok(tmp, sep_arr));
	else
		strcpy(result, tmp);
}

void parse_url(char url[], char host[], char path[])
{
	char tmp_url[MAXLEN];

	// get rid of fragments if case
	split(url, '#', tmp_url);

	// get rid of protocol
	if (strstr(tmp_url, "://")) {
		// here we only take the part of the url after the protocol
		strcpy(url, strstr(tmp_url, "://") + 3);
	}

	// now get the host
	split(url, '/', host);
	strcpy(tmp_url, url + strlen(host));

	// now check if we have a path and get it
	split(tmp_url, '?', path);
}

int is_malicious(char host[], char path[], char **blacklist, int nr_hosts)
{
	// check if host is blacklisted
	int i = 0;
	for (i = 0; i < nr_hosts; i++) {
		if (strstr(host, blacklist[i]))
			return 1;
	}

	// check for long hostnames
	if (strlen(host) > 31)
		return 1;

	// check for bad file extensions
	if (strstr(path, ".exe") || strstr(path, ".bin") || strstr(path, ".sh"))
		return 1;

	// if connecting to a specific port or credentials, maybe malicious
	if (strstr(host, "@") || strstr(host, ":"))
		return 1;

	// check for bad chars
	if (strstr(path, "~") && !strstr(path, ".htm"))
		return 1;

	// if only numbers and dots may be a malicious ip
	int no_numbers = 0;
	for (i = 0; i < (int)strlen(host); i++)
		if (host[i] >= '0' && host[i] <= '9')
			no_numbers++;
	if (no_numbers >= (float)(0.1 * strlen(host)))
		return 1;

	// check if 'com' is present multiple times
	char *p = strstr(host, "com");
	if (p)
		if (strstr(p + 3, "com"))
			return 1;

	// check for some weird strings
	if (strstr(path, "secur") || strstr(path, "paypal") ||
		strstr(path, "wp-admin"))
		return 1;

	return 0;
}

int task1(void)
{
	// load the urls file and the blacklist domains_database
	FILE *urls_file = fopen("../data/url_dataset/urls.in", "r");
	FILE *blacklist_file = fopen("../data/url_dataset/domains_database", "r");
	FILE *predictions_file = fopen("urls-predictions.out", "w");

	int i = 0;
	char **blacklist = malloc(SIZE * sizeof(char *));
	if (!blacklist) {
		printf("Malloc error\n");
		exit(-1);
	}
	for (i = 0; i < SIZE; i++) {
		char *tmp_v = calloc(MAXLEN, sizeof(char));
		if (!tmp_v) {
			printf("Malloc error\n");
			exit(-1);
		}
		blacklist[i] = tmp_v;
	}

	// read the contents of the blacklist
	int nr_hosts = 0;
	while (1) {
		fscanf(blacklist_file, " %s", blacklist[nr_hosts++]);
		if (feof(blacklist_file))
			break;
	}
	fclose(blacklist_file);

	// now get each url and check if it is malicious
	char url[MAXLEN], host[MAXLEN], path[MAXLEN];
	while (1) {
		fscanf(urls_file, " %s", url);
		if (feof(urls_file))
			break;

		// parse the url
		parse_url(url, host, path);

		//check if malicious
		if (is_malicious(host, path, blacklist, nr_hosts)) {
			fprintf(predictions_file, "1\n");
			strcpy(blacklist[nr_hosts++], host);
		} else {
			fprintf(predictions_file, "0\n");
		}
	}

	// close everything
	fclose(predictions_file);
	fclose(urls_file);

	// and free memory
	for (i = 0; i < SIZE; i++)
		free(blacklist[i]);
	free(blacklist);

	return 0;
}

void parse_traffic(char packet[], double *duration, double *payload_avg,
				   char src_ip[], char dst_ip[])
{
	int field_no = 0;
	*duration = 0;

    // now split the packet into parts
	char *field = strtok(packet, ":, ");
	while (field) {
		switch (field_no) {
		case SRC_IP:
			strcpy(src_ip, field);
			break;

		case DST_IP:
			strcpy(dst_ip, field);
			break;

		case FLOW_DURATION_DAYS:
			*duration += DAY_SECONDS * atof(field);
			break;

		case FLOW_DURATION_HOURS:
			*duration += HOUR_SECONDS * atof(field);
			break;

		case FLOW_DURATION_MINS:
			*duration += MINUTE_SECONDS * atof(field);
			break;

		case FLOW_DURATION_SECS:
			*duration += atof(field);
			break;

		case FLOW_PAYLOAD_AVG:
			*payload_avg = atof(field);
			break;

		default:
			break;
		}

		field = strtok(NULL, ":, ");
		field_no++;
	}
}

int is_malicious_traffic(char packet[])
{
	double duration, payload_avg;
	char src_ip[IP_MAXLEN], dst_ip[IP_MAXLEN];

	parse_traffic(packet, &duration, &payload_avg, src_ip, dst_ip);

    // check if payload is 0
	if (payload_avg == 0)
		return 0;

    // if broadcast, probably not malicious
	if (!strcmp(dst_ip, "255.255.255.255"))
		return 0;

    // check for duration
	if (duration > 1.0)
		return 1;

    // check for known signs of cryptominer
	if (payload_avg == 40.0)
		return 1;

	return 0;
}

int task2(void)
{
    // load the urls file and the blacklist domains_database
	FILE *traffic_file = fopen("../data/network_dataset/traffic.in", "r");
	FILE *predictions_file = fopen("traffic-predictions.out", "w");

    // now get each packet and check if it is malicious
	char packet[SIZE];
	char ip[IP_MAXLEN];
	fgets(packet, 300, traffic_file);
	while (1) {
		fgets(packet, 200, traffic_file);
		if (feof(traffic_file))
			break;

		//check if malicious
		int res = is_malicious_traffic(packet);
		if (res)
			fprintf(predictions_file, "1\n");
		else
			fprintf(predictions_file, "0\n");
	}

    // close everything
	fclose(predictions_file);
	fclose(traffic_file);

	return 0;
}

int main(void)
{
	task1();
	task2();

	return 0;
}
